Add iterator like functionality for DispatchKeySet

linux-jedi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44066 Add iterator like functionality for DispatchKeySet**

Add STL Input iterator to DispatchKeySet:
* Iterator is able to iterate from first not undefined DispatchKey
to NumDispatchKeys.
* Iterator is invalidated once underlying DispatchKeySet is invalidated

Note see http://www.cplusplus.com/reference/iterator/ for comparisons of
different iterators.

linux-jedi committed:
Add iterator like functionality for DispatchKeySet

Add STL Input iterator to DispatchKeySet:
* Iterator is able to iterate from first not undefined DispatchKey
to NumDispatchKeys.
* Iterator is invalidated once underlying DispatchKeySet is invalidated

Note see http://www.cplusplus.com/reference/iterator/ for comparisons of
different iterators.

[ghstack-poisoned]

ezyang commented on c10/core/DispatchKeySet.h:
```diff
@@ -120,6 +120,44 @@ class DispatchKeySet final {
+  // STL iterator for DispatchKeySet. Iterates from the first DispatchKey to the
+  // NumDispatchKey enum. The iterate is only invalidated by the destruction of
+  // the underlying DispatchKeySet as the iterator stores a pointer to the raw
+  // represenation of the DispatchKeySet.
```
AKA `std::list` style iterators, amirite? (Holding a pointer is pretty interesting, but it makes the iterators update if the input mutated so I guess that's a reasonable reason to do it that way.)

linux-jedi replied:
Ya that's correct, it's a list style iterator. None of the functions in `DispatchKeySet` actually mutate the set; instead, they return a new set with mutation, so the input representation shouldn't ever mutate.

ezyang commented on c10/core/DispatchKeySet.h:
```diff
@@ -120,6 +120,44 @@ class DispatchKeySet final {
+    using value_type = bool;
+    using difference_type = ptrdiff_t;
+    using pointer = const bool*;
+    using reference = const bool&;
```
These seem a bit questionable. What bool are you going to point to? The situation here is analogous to `std::vector`, where `reference` is a proxy class and the pointer is implementation-defined.

linux-jedi replied:
I wasn't really sure what the `pointer` should be (even if I change the `value_type` to `DispatchKey`). Right now, I'm taking a second pass that only iterates through `DispatchKey`s in the set, but I', still not sure what the pointer should be. My current understanding is that when the iterator is dereferenced a `DispatchKey` value should be returned. However, I'm not sure what a pointer would like in this case; the underlying data representation is a `uint64_t` that we convert to `DispatchKey` when the iterator is dereferenced. Any advice?

ezyang replied:
I'm not sure if you can omit these typedefs, but if you can, that would probably be what I would suggest.

linux-jedi commented on c10/core/DispatchKeySet.h:
```diff
@@ -120,6 +120,44 @@ class DispatchKeySet final {
+   public:
+    using self_type = iterator;
+    using iterator_category = std::input_iterator_tag;
+    using value_type = bool;
```
Okay that makes more sense. I think I misunderstood the functionality that we are looking for. Just to clarify, we want an iterator that iterates through only the `DispatchKeys` that are in a `DispatchKeySet`.

ezyang replied:
Yup!

linux-jedi committed:
Update on "Add iterator like functionality for DispatchKeySet"


Add STL Input iterator to DispatchKeySet:
* Iterator is able to iterate from first not undefined DispatchKey
to NumDispatchKeys.
* Iterator is invalidated once underlying DispatchKeySet is invalidated

Note see http://www.cplusplus.com/reference/iterator/ for comparisons of
different iterators.

[ghstack-poisoned]

linux-jedi commented:
Just pushed a revision so that iterator behaves as described in [above comments](https://github.com/pytorch/pytorch/pull/44066#discussion_r482681827)

ezyang commented on c10/core/DispatchKeySet.h:
```diff
@@ -120,6 +120,65 @@ class DispatchKeySet final {
+
+public:
+  // STL iterator for DispatchKeySet. Iterates through all DispatchKeys in the
+  // set. The iterator is only invalidated by the destruction ofthe underlying
```
nit: "of the"

ezyang commented on c10/core/DispatchKeySet.h:
```diff
@@ -120,6 +120,65 @@ class DispatchKeySet final {
+  class iterator {
+   public:
+    using self_type = iterator;
+    using iterator_category = std::input_iterator_tag;
```
Input iterator is probably good enough for all forseeable use cases we have, but realistically you probably actually have a bidirectional iterator here!

linux-jedi replied:
Agreed, with the current implementation it should be fairly easy to upgrade to a bidirectional iterator.

ezyang commented on c10/core/DispatchKeySet.h:
```diff
@@ -120,6 +120,65 @@ class DispatchKeySet final {
+
+      // If there are no keys, set to end iterator value
+      if (firstKeyIndex == std::numeric_limits<uint64_t>::max() ||
+          i_ >= static_cast<uint8_t>(DispatchKey::NumDispatchKeys)) {
```
`i_ >= static_cast<uint8_t>(DispatchKey::NumDispatchKeys)` is probably assert worthy behavior

ezyang commented on c10/test/core/DispatchKeySet_test.cpp:
```diff
@@ -53,3 +53,53 @@ TEST(DispatchKeySet, Full) {
+  uint8_t i = 0;
+
+  for (auto it = full_set.begin(); it != full_set.end(); ++it) {
+    i++;
```
It seems like it would also be a good idea to assert that `*it == static_cast<DispatchKey>(i)`

ezyang commented on c10/test/core/DispatchKeySet_test.cpp:
```diff
@@ -53,3 +53,53 @@ TEST(DispatchKeySet, Full) {
+  for (auto it = full_set.begin(); it != full_set.end(); ++it) {
+    i++;
+  }
+  ASSERT_EQ(i, static_cast<uint8_t>(DispatchKey::NumDispatchKeys) - 1);
```
This looks pretty odd. If there are three dispatch keys, if I iterate through each of them, incrementing i once per, then I'd expect i == 3! (Not 2!) Maybe this is right, but it smells like an off-by-one error to me.

linux-jedi replied:
I don't believe it is. `NumDispatchKeys` is an actual DispatchKey enum (and has a uint8_t conversion) that always comes last. Additionally, the DispatchKey that has a value of zero, represents undefined or an empty key set. Consequentially, If there are three dispatch keys, `NumDispatchKeys` will have a unit8_t value of 4.

ezyang replied:
You're right, zero is not included, so -1.

ezyang commented on c10/test/core/DispatchKeySet_test.cpp:
```diff
@@ -53,3 +53,53 @@ TEST(DispatchKeySet, Full) {
     ASSERT_TRUE(full.has(tid));
   }
 }
```
I'd like to see a test that allocates a dispatch key set with, say, three elements over it, iterates over it, and then shows that what you iterated over is exactly what was in the dispatch key set. All the tests here don't look like they would identify the class of bugs that a test like this would identify.

ezyang reviewed (changes requested):
Looking for more tests

linux-jedi committed:
Update on "Add iterator like functionality for DispatchKeySet"


Add STL Input iterator to DispatchKeySet:
* Iterator is able to iterate from first not undefined DispatchKey
to NumDispatchKeys.
* Iterator is invalidated once underlying DispatchKeySet is invalidated

Note see http://www.cplusplus.com/reference/iterator/ for comparisons of
different iterators.

[ghstack-poisoned]

linux-jedi commented:
Added tests:
* check that iterator cannot overflow past end position
* iterator actually iterates over specific keys in sparsely populated keyset

linux-jedi committed:
Update on "Add iterator like functionality for DispatchKeySet"


Add STL Input iterator to DispatchKeySet:
* Iterator is able to iterate from first not undefined DispatchKey
to NumDispatchKeys.
* Iterator is invalidated once underlying DispatchKeySet is invalidated

Note see http://www.cplusplus.com/reference/iterator/ for comparisons of
different iterators.

[ghstack-poisoned]

ailzhang reviewed (approved):
LGTM! Thanks!
